boards: nxp: adjusting the scope of boot-related build flags

ZhaoxiangJin opened the pull request:
* For the RT10xx and RT11xx families, the compile flag ``BOARD_FLASH_SIZE``, originally defined in
    ``boards/nxp/mimxrt10xx_evk/CMakeLists.txt`` and ``boards/nxp/mimxrt11xx_evk/CMakeLists.txt``, is
    used only by the HAL header ``fsl_flexspi_nor_boot.h``, which is included by
    :zephyr_file:`soc/nxp/imxrt/imxrt10xx/soc.c` and :zephyr_file:`soc/nxp/imxrt/imxrt11xx/soc.c`.
    To avoid potential collisions with other global flags, the macro is now defined at the SoC layer
    using ``zephyr_library_compile_definitions()`` in :zephyr_file:`soc/nxp/imxrt/imxrt10xx/CMakeLists.txt`
    and :zephyr_file:`soc/nxp/imxrt/imxrt11xx/CMakeLists.txt`. This change has been applied to all
    RTxxxx boards.

  * For the RTxxx family, the compile flag ``BOARD_FLASH_SIZE``, originally defined in
    ``boards/nxp/mimxrtxxx_evk/CMakeLists.txt``, is not used in the Zephyr tree and has
    therefore been removed from all RTxxx board CMakeLists.txt files.

  * For the RTxxx family, the compile flag ``BOOT_HEADER_ENABLE``, previously defined in
    ``boards/nxp/mimxrtxxx_evk/CMakeLists.txt`` and used in ``boards/nxp/rtxxx/<boot_header>.c``,
    has been replaced by a Kconfig option. Consequently, the line
    ``zephyr_compile_definitions(BOOT_HEADER_ENABLE=1)`` has been removed from the RTxxx board
    CMakeLists.txt files.

  * Removed compile flag ``BOOT_HEADER_ENABLE`` definition from :zephyr_file:`boards/nxp/rd_rw612_bga/CMakeLists.txt`,
    as it is not used in the Zephyr tree.

  * Originally, the compile flags ``XIP_BOOT_HEADER_ENABLE`` and ``XIP_BOOT_HEADER_DCD_ENABLE`` were
    used in ``boards/nxp/rt1xxx/<boot_header>.c``. These flags have been converted to Kconfig options
    across NXP RTxxxx evaluation boards, allowing boot-header configuration via the Kconfig build system
    instead of compile-time defines. Consequently, we removed ``zephyr_compile_definitions(XIP_BOOT_HEADER_ENABLE=1)``
    and ``zephyr_compile_definitions(XIP_BOOT_HEADER_DCD_ENABLE=1)`` from the RTxxxx board-level CMakeLists.txt files.
    Because these macros are also required by ``hal_nxp/rt10xx/fsl_flexspi_nor_boot.h`` and
    ``hal_nxp/rt11xx/fsl_flexspi_nor_boot.h``, they were added to the corresponding SoC-layer CMakeLists.txt files
    using ``zephyr_library_compile_definitions()`` to limit their scope.

rework https://github.com/zephyrproject-rtos/zephyr/pull/96902

ZhaoxiangJin commented:
hello @hakehuang, could you please help to run regression test base on this PR? thanks.

hakehuang commented:
> hello @hakehuang, could you please help to run regression test base on this PR? thanks.

will test and feedback

ZhaoxiangJin committed:
boards: nxp: convert 'XIP_BOOT_HEADER_DCD_ENABLE' to Kconfig options

1. This commit converts the 'XIP_BOOT_HEADER_DCD_ENABLE' preprocessor
define to Kconfig options across NXP i.MX RTxxxx evaluation boards.
The changes enable flexible boot header configuration through the
Kconfig build system instead of compile-time defines, improving
configurability and maintainability.

2. Now 'XIP_BOOT_HEADER_DCD_ENABLE' flag is only used in
hal_nxp/rtxxxx/fsl_flexspi_nor_boot.h which is included by soc.c,
therefore, this flag should be passed in the rtxxxx/SoC layer CMake
using zephyr_library_compile_definitions(). and can be removed from
each board CMakeLists.txt.

Note: if out-of-tree files still need to use this flag, please add
it in your own CMakeLists.txt.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

MaochenWang1 commented on boards/nxp/frdm_rw612/CMakeLists.txt:
```diff
@@ -23,12 +22,3 @@ if(CONFIG_DT_HAS_NXP_ENET_MAC_ENABLED AND ${xtal32_status})
     "between the ethernet PHY and the external oscillator")
 endif()
-
-# Set TX power limit file to override the default one
```
could we keep this part unchanged?
It's an example showed to customer to add their own WIFI_BT_TX_PWR_LIMITS_OVERRIDE on different boards.

ZhaoxiangJin commented:
cc @lucien-nxp

nordicjm reviewed (dismissed):
great, thanks

nordicjm commented:
Although having looked into the macro function, isn't that just doing the exact same thing as zephyr_compile_definitions is and adding things globally?

ZhaoxiangJin commented:
> Although having looked into the macro function, isn't that just doing the exact same thing as zephyr_compile_definitions is and adding things globally?

yes, this function also add flags to global targets, changed to use `zephyr_library_compile_definitions`.

ZhaoxiangJin commented:
> > 1. some definitions are not used in zephyr and hal_nxp, so remove them, for examples: [commit1](https://github.com/zephyrproject-rtos/zephyr/pull/100252/commits/629e76398f17693281e1f02e1be29dad6cbc89e3).
> 
> mcxn947 do have flash boot option, why remove it?

Actually through a global search, I found that this flag is not used by mcxn947. I don't know why it was added, perhaps @mmahadevan108 and @DerekSnell know something.

> > 2. some definitions are not used in hal_nxp only, so move them to hal_nxp cmake, for examples: [commit2](https://github.com/zephyrproject-rtos/zephyr/pull/100252/commits/757b8e9b1afe4fbe1c158005dd4170e6543716c1), [commit3](https://github.com/zephyrproject-rtos/zephyr/pull/100252/commits/24a60d462080031dec289f34dd0a3eddbc4b6ebe) and [commit4](https://github.com/zephyrproject-rtos/zephyr/pull/100252/commits/cc16ec98bef9beb43ea02ac754479c493e4b8d07).
> 
> I am ok with [commit2](https://github.com/zephyrproject-rtos/zephyr/pull/100252/commits/757b8e9b1afe4fbe1c158005dd4170e6543716c1)
> 
> but see below, I do not believe the BOARD_FLASH_SIZE bind with SOC is correct. they should always go with the platform.
> 
> ```
> if(CONFIG_SOC_SERIES_IMXRT10XX OR CONFIG_SOC_SERIES_IMXRT11XX)
>   if(CONFIG_NXP_IMXRT_BOOT_HEADER AND CONFIG_BOOT_FLEXSPI_NOR)
>     mcux_add_macro(BOARD_FLASH_SIZE=CONFIG_FLASH_SIZE*1024)
>   endif()
> ```

Yes, you are correct, but after analysis, I found that this flag should be added in the soc level cmake (reasons can be found in commit message), the PR has been updated ([commit3](https://github.com/zephyrproject-rtos/zephyr/pull/100252/commits/91e40fa60280c8cad8be2fbbeb9d8689a410586d)), please review, thanks.

> and for
> 
> ```
> if(CONFIG_SOC_SERIES_IMXRT7XX AND CONFIG_IMX_USDHC)
>   mcux_add_macro(FSL_SDK_ENABLE_DRIVER_CACHE_CONTROL=1)
> endif()
> ```
> 
> the Driver cache control is required for RT7xxx, it is not only for hal, but for shim driver as well. My concerns are thinking about if we are moving to riscv/rust solution those need a quick integration with Zephyr, and we need skip our current arm based SDK. then it is the necessary to move Marcos from hal to Zephyr native, not reverse.

This flag is added [here](https://github.com/zephyrproject-rtos/zephyr/pull/91316/files#diff-e85c6c4b9eddde2584dbea9a2dcbaacbfa81151e314c4d405a1b8e1ab48e8d17R31), after through a global search, I found that this flag is only used in hal_nxp fsl_usdhc.c and fsl_enet.c, then I think it is better add it in glue cmake. 

> > 3. some definitions are used in zephyr only, so convert to use `zephyr_library_compile_definitions`, for examples: [commit8](https://github.com/zephyrproject-rtos/zephyr/pull/100252/commits/438aef826e7e8e03f3194354bc2b38d56b8c2295).
> 
> I would encourage to reuse the zephyr defines as possible, instead of defining another one, and translate them in hal, remember the policy is `Zephyr first`.

`zephyr_library_compile_definitions` is a zephyr function, it is better than `zephyr_compile_definitions` because `zephyr_library_compile_definitions` passes the flag only to the target that requires it.

> > 4. some definitions are used in zephyr and hal_nxp, but according [Jmie](https://github.com/zephyrproject-rtos/zephyr/pull/95705#discussion_r2343840196)'s suggestion, uses  `zephyr_library_compile_definitions` is a better way, so in zephyr side we use `zephyr_library_compile_definitions`, in hal_nxp side we use `mcux_add_macro `, for examples: [commit5](https://github.com/zephyrproject-rtos/zephyr/pull/100252/commits/4225b7b43c601846d5fb9e273a9b8c4ce569884f), [commit6](https://github.com/zephyrproject-rtos/zephyr/pull/100252/commits/e35595f98ecf1791e83db2c209cd976962c5ff42) and [commit7](https://github.com/zephyrproject-rtos/zephyr/pull/100252/commits/235870428e98b9d4b53d591f687497657e2cc59c).
> 
> I would suggest that we clarify with [Jmie] on this, as those are required for NXP soc and its hal, so a global Marco is good in this case. Adding a translation layer is very prune to error. we can just add a comment in those cmakefile to clarify this.

Currently most flags are passed in the soc level cmake, could you please help to review again? then we can continue discuss, thanks.

DerekSnell commented:
> > > 1. some definitions are not used in zephyr and hal_nxp, so remove them, for examples: [commit1](https://github.com/zephyrproject-rtos/zephyr/pull/100252/commits/629e76398f17693281e1f02e1be29dad6cbc89e3).
> > 
> > 
> > mcxn947 do have flash boot option, why remove it?
> 
> Actually through a global search, I found that this flag is not used by mcxn947. I don't know why it was added, perhaps @mmahadevan108 and @DerekSnell know something.
> 
Hi @ZhaoxiangJin and @hakehuang ,
I see you are right, and the flag `XIP_BOOT_HEADER_ENABLE` does not need to be set for the MCX devices.  Like you, I do not see this flag used anywhere in the HAL for those devices.  And I just tested on the board variant `frdm_mcxn947//cpu0/qspi` and it boots without that flag.

I see that flag is used on the RT1xxx devices, so I suspect it was copied from them when QSPI boot was enabled on this platform.  But we can remove it.  Thanks

ZhaoxiangJin committed:
doc: releases: add migration guide for NXP compile flag scope changes

Document the scope changes for NXP compile flags that were
moved from global to local scope in PR #100252. Applications
that depended on these flags being globally available may need
to be updated.

The following flags have been moved to use
zephyr_library_compile_definitions():
- BOARD_FLASH_SIZE: moved to SoC layer CMake for RT10xx and RT11xx
- XIP_BOOT_HEADER_ENABLE: scoped to boards CMake or SoC layer as needed
- BOOT_HEADER_ENABLE: scoped to boards CMake, removed from RW61x boards
- XIP_BOOT_HEADER_DCD_ENABLE: scoped to boards CMake or SoC layer as needed

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin committed:
soc: nxp: scope boot-header compile defs to soc library

Create a dedicated Zephyr library for the imxrt11xx and
imxrt10xx SoC in CMakeLists.txt and move soc.c into it
(zephyr_library() + zephyr_library_sources(soc.c)).

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin committed:
boards: nxp: convert 'BOOT_HEADER_ENABLE' to Kconfig options

1. This commit converts the 'BOOT_HEADER_ENABLE' preprocessor
define to Kconfig options across NXP i.MX RTxxx evaluation boards.
The changes enable flexible boot header configuration through the
Kconfig build system instead of compile-time defines, improving
configurability and maintainability.

2. Now this flag is not used in Zephyr tree, so removed
'zephyr_compile_definitions(BOOT_HEADER_ENABLE=1)' from the
CMakeLists.txt files of the following boards:
- mimxrt595_evk
- mimxrt685_evk
- mimxrt700_evk

Note: for out-of-tree projects that still require this flag,
users can define it in their own CMakeLists.txt or prj.conf files.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin committed:
boards: nxp: convert 'XIP_BOOT_HEADER_ENABLE' to Kconfig options

1. This commit converts the 'XIP_BOOT_HEADER_ENABLE' preprocessor
define to Kconfig options across NXP i.MX RTxxxx evaluation boards.
The changes enable flexible boot header configuration through the
Kconfig build system instead of compile-time defines, improving
configurability and maintainability.

2. Now 'XIP_BOOT_HEADER_ENABLE' flag is only used in
hal_nxp/rt11xx/fsl_flexspi_nor_boot.h which is included by soc.c,
therefore, this flag should be passed in the rt11xx/SoC layer CMake
using zephyr_library_compile_definitions(). and can be removed from
each board CMakeLists.txt.

Note: if out-of-tree files still need to use this flag, please add
it in your own CMakeLists.txt.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin committed:
boards: nxp: rd_rw612_bga: remove unused build flag

'BOOT_HEADER_ENABLE' is not used in RW61x build tree,
so should remove it from RW61x board CMakeLists.txt.

Note: if out-of-tree files still need to use this flag,
please add it in your own CMakeLists.txt.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin committed:
boards: nxp: remove 'BOARD_FLASH_SIZE' for RTxxx boards

Flag 'BOARD_FLASH_SIZE' is not used for RTxxx, so remove it
from mimxrt595_evk, mimxrt685_evk, and mimxrt700_evk board
CMakeLists.txt files.

Normally, 'BOARD_FLASH_SIZE' should not be used elsewhere.
If an out-of-tree user uses this flag elsewhere, please define
the flag yourself.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin committed:
boards: nxp: move 'BOARD_FLASH_SIZE' to SoC layer cmake

For RT10xx and RT11xx, flag 'BOARD_FLASH_SIZE' is only used
in hal_nxp fsl_flexspi_nor_boot.h which is included by soc.c,
therefore, this flag should be passed in the SoC layer CMake
using zephyr_library_compile_definitions() to limit its scope.

Normally, 'BOARD_FLASH_SIZE' should not be used elsewhere.
If an out-of-tree user uses this flag elsewhere, please define
the flag yourself.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin commented on soc/nxp/imxrt/imxrt10xx/CMakeLists.txt:
```diff
@@ -22,6 +22,12 @@ if(CONFIG_MEMC_MCUX_FLEXSPI)
 
+if(CONFIG_NXP_IMXRT_BOOT_HEADER)
+  if(CONFIG_BOOT_FLEXSPI_NOR)
+    zephyr_library_compile_definitions(BOARD_FLASH_SIZE=CONFIG_FLASH_SIZE*1024)
```
fixed actually, see https://github.com/zephyrproject-rtos/zephyr/pull/100252/changes/c484e026bacbed2ff72590abb209e514f0ac4693.
<img width="1710" height="283" alt="image" src="https://github.com/user-attachments/assets/2315b390-8426-415e-b5b3-b38733e4b37d" />

dnltz commented:
@mmahadevan108 can you please take a look at this PR?
